### PR TITLE
MB-2464 [JM] api test proof of service upload

### PR DESCRIPTION
**Pull request by jacquelineIO**
## Description

Added new endpoint to `prime-api-client` to create an upload. 

Bypassing using `json` because I'm not sure how to write a json file to represent a file. Decided to go a different route and just use flags to pull in the `paymentRequestID` and the `filename` to be uploaded which is not a `json` file like with other endpoints. 

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup
```
make db_dev_reset db_dev_e2e_populate
rm bin/prime-api-client && make bin/prime-api-client 
```

To run the upload command you follow the sequence:
1. Fetch MTOs, to find the MTO ID that you will be working with (and also to find the service items IDs if you are adding those to the payment request)
2. Create a payment request and then use the payment request ID that is returned
3. Create a payment request upload

(Copying heavily from PR #3932 in the setup section)
### Testing Dev

1. First run `make db_dev_e2e_populate` to ensure your database is setup with the E2E data.
2. Next, save this example JSON payload (which references E2E UUIDs) to a file called `pr_payload_dev.json`:
```
{
    "isFinal": false,
    "moveTaskOrderID": "da3f34cc-fb94-4e0b-1c90-ba3333cb7791",
    "serviceItems": [
        {
            "id": "8a625314-1922-4987-93c5-a62c0d13f053",
            "params": [
                {
                    "key": "WeightEstimated",
                    "value": "3254"
                },
                {
                    "key": "RequestedPickupDate",
                    "value": "2020-04-20"
                }
            ]
        }
    ]
}
```
3. Ensure your server is running: `make server_run`
4. Run this command to fetch the "available to prime" MTOs (there's just one of those in E2E data at this time) and notice that there are no payment requests for it yet in the returned JSON: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
5. Now let's create a payment request using our payload above.  There are two different ways to do this:
    - With the `--filename` flag: `go run ./cmd/prime-api-client --insecure create-payment-request --filename pr_payload_dev.json | jq`
    - From stdin: `go run ./cmd/prime-api-client --insecure create-payment-request - < pr_payload_dev.json | jq`
6. Fetch the MTOs again and verify the payment request(s) you created are there: `go run ./cmd/prime-api-client --insecure fetch-mtos | jq`
7. Create an upload for the payment request using the payment request ID
`go run ./cmd/prime-api-client --insecure create-payment-request-upload --paymentRequestID "fe3bfc25-0c94-4783-80f8-132c73f4cea1" --filename "/Users/jacquelinemckinney/Documents/mymove/pcs orders/dd1614_pcs.pdf" | jq`

### Testing Experimental (optional)

**NOTE: You must have a CAC to do this part!**

1. Make a copy of the JSON payload above and save it to `pr_payload_exp.json`
2. See what "available to prime" MTOs are out there (you will prompted for your CAC PIN): `go run ./cmd/prime-api-client --cac --hostname api.experimental.move.mil --port 443 fetch-mtos | jq`
3. Edit the `pr_payload_exp.json` file and replace the two UUIDs in it with a `moveTaskOrderID` value and `mtoServiceItems` ID value from the output of step 2. 
4. Create the payment request: `go run ./cmd/prime-api-client --cac --hostname api.experimental.move.mil --port 443 create-payment-request --filename pr_payload_exp.json | jq`
5. Fetch the MTOs again and verify the payment request(s) you created are there: `go run ./cmd/prime-api-client --cac --hostname api.experimental.move.mil --port 443 fetch-mtos | jq`
6. Create an upload for the payment request using the payment request ID
`go run ./cmd/prime-api-client  --cac --hostname api.experimental.move.mil --port 443 create-payment-request-upload --paymentRequestID "fe3bfc25-0c94-4783-80f8-132c73f4cea1" --filename "/Users/jacquelinemckinney/Documents/mymove/pcs orders/dd1614_pcs.pdf" | jq`
7. Verify file is in AWS S3 bucket by going to the S3 website or by running AWS CLI, if running AWS CLI to check run the following: `aws s3 ls --recursive s3://transcom-ppp-app-experimental-us-west-2/app/app/payment-request-uploads`
8. Once verified to remove file run: 
`$ aws s3 rm s3://transcom-ppp-app-experimental-us-west-2/app/app/payment-request-uploads/mto-5d4b25bb-eb04-4c03-9a81-ee0398cb779e/payment-request-80ff7576-b3f9-4812-85bd-6bc14f02d862/timestamp-2020-05-04 20:52:50.236046446 +0000 UTC m=+830.171653999` (can't do recursive remove based on user when not in local dev)

### Testing Staging (follow steps for experimental)

**Testing in the staging environment should be similar, but uses `api.staging.move.mil` instead.
transcom-ppp-app-staging-us-west-2**

Verify file is in AWS S3 bucket by going to the S3 website or by running AWS CLI, if running AWS CLI to check run the following: `aws s3 ls --recursive s3://transcom-ppp-app-staging-us-west-2/app/app/payment-request-uploads`

Once verified to remove file run: 
`$ aws s3 rm  s3://transcom-ppp-app-staging-us-west-2/app/app/payment-request-uploads/mto-5d4b25bb-eb04-4c03-9a81-ee0398cb779e/payment-request-80ff7576-b3f9-4812-85bd-6bc14f02d862/timestamp-2020-05-04 20:52:50.236046446 +0000 UTC m=+830.171653999`

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/blob/master/docs/database/migrate-the-database.md#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2464) for this change

## Screenshots
n/a

### CODE REVIEW

File: cmd/prime-api-client/create_payment_request_upload.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+	"github.com/transcom/mymove/pkg/gen/primeclient/uploads"
+)
+
+// initCreatePaymentRequestUploadFlags initializes flags.
+func initCreatePaymentRequestUploadFlags(flag *pflag.FlagSet) {
+	flag.String(FilenameFlag, "", "Path to the upload file for create-payment-request-upload payload")
+	flag.String(PaymentRequestID, "", "Payment Request ID to upload the proof of service document to")
+
+	flag.SortFlags = false
```
Comment (gidjin) [inline cmd/prime-api-client/create_payment_request_upload.go:24]: I'm curious why do you need this set to false for?

File: cmd/prime-api-client/create_payment_request_upload.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	if errCreateClient != nil {
+		return errCreateClient
+	}
+
+	// Defer closing the store until after the API call has completed
+	if cacStore != nil {
+		defer cacStore.Close()
+	}
+
+	// Decode json from file that was passed into create-payment-request-upload
```
Comment (gidjin) [inline cmd/prime-api-client/create_payment_request_upload.go:71]: Looks like this comment is leftover from a previous iteration?

File: cmd/prime-api-client/create_payment_request_upload.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	filename := v.GetString(FilenameFlag)
+	if filename == "" {
+		return fmt.Errorf("failed to open filename: %s", filename)
+	}
```
Comment (gidjin) [inline cmd/prime-api-client/create_payment_request_upload.go:75]: This error check feels redundant. Is it not enough to check if filename is provided in `checkCreatePaymentRequestUploadConfig`?

File: cmd/prime-api-client/create_payment_request_upload.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	paymentRequestID := v.GetString(PaymentRequestID)
+	if paymentRequestID == "" {
+		return fmt.Errorf("paymentRequestID required: %s", paymentRequestID)
+	}
```
Comment (gidjin) [inline cmd/prime-api-client/create_payment_request_upload.go:80]: I'd probably check `PaymentRequestID` in `checkCreatePaymentRequestUploadConfig` along with `FilenameFlag` or maybe check both flags here. Either way is fine but I would like the checks to be together.

**Comment by killrob234** [comment]
Can some tell me how to make this work with my phone

**Pull request merged**